pkg/input/arm/eval.go

```diff
@@ -22,23 +22,20 @@ import (
 // references to implementations of functions, and the other fields exist to
 // pass data to certain function impls other than what is already in their args.
 type EvaluationContext struct {
-	discoveredResourceSet map[string]struct{}
-	variables             map[string]interface{}
-	funcs                 map[string]armFnImpl
+	DiscoveredResourceSet map[string]struct{}
+	Variables             map[string]interface{}
+	BuiltinFunctions      map[string]Function
 }
 
-func NewEvaluationContext(discoveredResourceSet map[string]struct{}, variables map[string]interface{}) *EvaluationContext {
-	evalCtx := &EvaluationContext{
-		discoveredResourceSet: discoveredResourceSet,
-		variables:             variables,
-		funcs: map[string]armFnImpl{
-			"concat":        concatImpl,
-			"resourceGroup": resourceGroupImpl,
-		},
+type Function func(e *EvaluationContext, args ...interface{}) (interface{}, error)
```

Comment:
This is all a bit speculative--

In general, having the same shape for these allows you to manipulate them a bit more easy, e.g. have helpers like `func memoize(f Function) Function`.

On the other hand, it _is_ useful to see which function depend on additional context.  it's a bit of a trade-off though, does a function like `variables()` depend on `EvaluationContext` or do we go even more granular and depend on `Variables` only?

Tying them to the evaluation pass seems sensible, but in that case we'd want to propagate that type to the context, and I'm not really sure if Go's generics will make that nice...

As a middle way, I'll:
- update the function implementations to not take any context if they don't have to
- keep the `Function` type like this because it accurately describes "everything" a function can do which is useful